### ISSUE TITLE
[performance] Agent improvements

### ISSUE THREAD

**Issue by srinivasupadhya** [gocd/gocd#1098]
Improve Agent Server communication to reduce the problem of too many agents pinging the server at same time during start up & too many requests (1 ping + 1 work request = 2 requests / sec / agent) from agents.
- Thundering herd - implement exponential backoff while registering so that Server is not swamped with requests.
- Websocket - move from polling to server push model

**Comment by greenmoss** [comment]
+1 on this! Rationale: every pipeline stage that is pending, must wait for an agent to check in. It seems the polling time is about one minute (not sure about this). So for a pipeline with 5 stages, the polling overhead can add up to 5 excess minutes to the pipeline's completion time. It would be great to get rid of that execution time overhead!

**Comment by kmugrage** [comment]
+1 on the Websocket ( or similar ) change in model. The delay between a material change whether code or another pipeline is quickly becoming my biggest user experience issue. It's quite painful when modelling new stuff.

**Comment by lenucksi** [comment]
Especially together with the configuration saving/validation improvements mentioned in #1498 this could really help to mitigate load and stability issues on large Go servers.

**Comment by xli** [comment]
Spiking on websocket solution, talk to me if you did something related :)

**Comment by ketan** [comment]
If you are :+1

**Comment by xli** [comment]
My plan
--------------------

* Upgrade Spring to 4.3.3 (latest):  not full upgrade for spike, just do enough work for spike.
* Use Spring 4's WebSocket support to implement server side communication.
* Use Jetty's WebSocket client for Agent implementation.

Other options in consideration:
* Bypass Spring to use Jetty's WebSocket. I feel there maybe headache of how to make it work with existing stuff that is integrated with Spring. So I think it is better to impl in Spring framework if we impl it in Java, at least try using Spring framework's websocket first.
* Rails websocket: Rails 5 will have websocket support. For Rails 4, there are some gems support it (em-websocket, websocket-rails, tubesock), but no active maintainer. So I don't feel strong confidence to use any existing Websocket impl in Rails. Plus I think most of people will choose to support Rails' ActionCable for websocket. Hence most of other websocket impl on rails will die eventually. 
* Ruby websocket server through rack: https://github.com/faye/faye-websocket-ruby (I found this today, seems good, will do more research on it later)

My spike work is pushed to: https://github.com/xli/gocd/tree/websocketspike

Progress
---------------
Upgrading Spring for spike has more work than I thought, especially the HttpClient part, Spring upgraded to use HttpClient 4, which is whole new impl comparing to HttpClient 3 we're using. Hence I need rewrite most stuff related to HttpClient on Agent.

So far, major Spring upgrade work for the spike is done (basic build function works in development environment), will start to implement Agent Server communication with Spring websocket. 
(cc @ketan @arvindsv @jyotisingh @zabil)

**Comment by ketan** [comment]
In the past I've seen issues with most libs that ultimately depend on eventmachine (running on jruby).

Last I checked — https://github.com/ngauthier/tubesock was a good alternative (although I did not quite use it, but it's pure ruby and so are its dependencies).

As regards "no maintainer" I believe most of the websocket APIs have matured enough to not need regular maintainance (atleast on MRI, jruby is another can of worms)

I'd probably see if doing this in ruby/rails is any easier :)

**Comment by ketan** [comment]
Also, is using the rack hijack "API" an alternative? http://old.blog.phusion.nl/2013/01/23/the-new-rack-socket-hijacking-api/

**Comment by xli** [comment]
faye-websocket-ruby can be integrated by rack hijack "API", and it is actively maintained, so far I found the best option for Ruby

**Comment by xli** [comment]
Pushed latest spike to https://github.com/xli/gocd/tree/websocketspike.
Converted Agent ping and retrieve work part into WebSocket communication on top of Spring 4 and Jetty WebSocket frameworks. 
 It would be great if someone can check it out and give me some feedback about it. (cc @ketan @arvindsv @jyotisingh @zabil)

Next: I'll do a quick try out with faye-websocket-ruby, then compare it with Spring&Jetty version.

**Comment by ketan** [comment]
Another spike at https://github.com/ketan/gocd/tree/websocket-spike - seems fairly straightforward using jetty websocket client and server — there was no need to upgrade any libraries.

I did a quick lookup for using rails, but unfortunately jruby-rack does not yet support the rack-hijack api so any ruby based websocket solutions are out the door.

**Comment by xli** [comment]
@ketan Thanks, it looks a lot simpler to use jetty websocket servelet comparing to spring websocket handler

**Comment by xli** [comment]
Progress update:

Spike is done. 

I started implement WebSocket communication between Agent and Server implementation:
* Use Jetty websocket in both client and server side
* Use JSON to encode and decode message between server and agent.
* Change server side push to deliver new job to agent.
* Agent will still ping server with agent runtime info periodically for:
  * keep connection alive.
  * robustness
* Will add a toggle (system environment property) in Agent side for switching to use websocket communication.
  * This toggle is designed for testing and continuous improvement before we can get ride of current spring RPC implementation.

My impl will be pushed https://github.com/xli/gocd/tree/jetty.websocket progressively

**Comment by xli** [comment]
Moved JSON format out of scope, because BuildWork has deep and complex object tree, use Java Serialization for first round impl.

**Comment by arvindsv** [comment]
I've done most of the work for it [here](https://github.com/arvindsv/gocd/commits/spike_agent_to_server_communication_using_json). Maybe once you're done, I can re-do it for websockets.

**Comment by xli** [comment]
@arvindsv sure,  I tried yesterday with Gson, hope to have a simple version works, but failed. So I think we probably can just copy your implementation later.

**Comment by xli** [comment]
progress and implementation details: https://github.com/xli/gocd/wiki/WebSocket-communication

**Comment by arvindsv** [comment]
Cool. I have some thoughts/questions:

1. You mention "reregistering" here:
> 
> How unregistered and disabled agents are handled on server?
> 
> Server always accepts agent websocket session no matter what agent status is when agent pings server. Server will send reregister message to agent if it found agent was not registered during the process of matching agents with jobs (similar with current behavior). We don't handle it when websocket session is opened, because agent may become unregistered status if user deleted agent from agent list. It is more robust way to handle unregistered agents. However, we won't do anything to disabled agents.

    My thought is, do we really need the complicated SSL certificate management we do today? I suppose that's what [you mean](https://github.com/xli/gocd/commit/d412dc203ac0583826f9062fe9254c2566c4a63d#diff-4aa6ab956d8b03c9c3fd4a4f958d21a6R285) by reregister. What is it giving us? Maybe more of a question for @jyotisingh, @ketan, @zabil, @maheshp to think about. What happens if we get rid of them and use the usual endpoint security?

2. You mention async:

    > Why Server sends message to Agent in asynchronous mode?

    Need to be careful that all cases are handled.

3. In this question:

    > What happened when agent lost websocket connection and is building something?
    > 
    > Agent reconnect to server building status, and it is still building the original job, if agent didn't restart.

    So, no reschedule and everything continues as normal, right?

    > If Agent restarted, agent will connect to server with idle status, which will update server side agent status.

    This will lead to a [reschedule](https://github.com/gocd/gocd/blob/ada9b36174cc069c860e7fa032cbf2857f135cfb/server/src/com/thoughtworks/go/server/service/ScheduleService.java#L597), I guess? We should consider putting a message in the console log of that job. Today, there is no message. Another agent just starts.

4. Websocket error: "What we do for handling websocket error?"

    Depends on the kind of error, I guess.

**Comment by xli** [comment]
@arvindsv I'll update the wiki page basing on your questions.

And here is more specific answers to your questions:

> In this question:

> What happened when agent lost websocket connection and is building something?

> Agent reconnect to server building status, and it is still building the original job, if agent didn't restart.
> So, no reschedule and everything continues as normal, right?

I tested this case, and found out ScheduleService#rescheduleHungJobs (about 5 mins) will reschedule job if agent didn't reconnect (which means it is lost contact status)

> If Agent restarted, agent will connect to server with idle status, which will update server side agent status.
> This will lead to a reschedule, I guess? We should consider putting a message in the console log of that job. Today, there is no message. Another agent just starts.

Yes, it will be rescheduled when we assign a new job to the reconnected idle agent (or for any reason, the agent status is changed to idle).

> Websocket error: "What we do for handling websocket error?"
> Depends on the kind of error, I guess.

Good point, I need write it down instead of assuming everyone will have similar thought process to go through every cases and come out same result with me.
As it's too much stuff, I updated the question, please click https://github.com/xli/gocd/wiki/WebSocket-communication#what-we-do-for-handling-websocket-error-on-server to see details

**Comment by FredrikWendt** [comment]
Do you plan to make available a pre-release version where one can start using this?

**Comment by garthk** [comment]
We're running out of August. Is this effort dead in the water?

**Comment by arvindsv** [comment]
Talked to @zabil. The websocket work is in and merged, but is an agent-side flag that needs to be turned on.  We tried a couple of months ago to enable it by default on some of our servers and found a couple of issues, including [this one](https://github.com/gocd/gocd/issues/2263).

Since that time, some issues have been [fixed](https://github.com/gocd/gocd/issues/2249) and Jetty has been upgraded. So, we need to try again to dogfood this change - turning on the flag by default, so that everyone uses websockets - and see if it uncovers any issues.

**Comment by garthk** [comment]
I'll give it a shot; how do I enable it?

**Comment by zabil** [comment]
You can turn it on by using the system following property
```
-Dgo.agent.websocket.enabled=true
```

While starting the agent. Just append the property to the environment variable ```GO_AGENT_SYSTEM_PROPERTIES``` and restart the agent.

Please note that the websocket functionality stopped working from 16.7+ releases. We haven't had a chance to fix this problem yet.

**Comment by garthk** [comment]
Aah. Ok, I'll wait.

**Comment by ciotlosm** [comment]
Any update on this? It would bring benefit to users having multiple stages that wait for agents. The seconds in the cycle add up quickly, and the poll model also adds stress to the server.

**Comment by arvindsv** [comment]
We recently talked about trying to turn this on by default. When we tried it a little while ago, the server ran out of memory. We found that we had to upgrade Jetty because they fixed some issues recently. I think @jyotisingh might be the best person to talk about the latest.

**Comment by FredrikWendt** [comment]
Would be interesting to hear if you've ran this again with acceptable results. If so, we'd like to give this a try. When doing so, anything in particular should we monitor or take a look at?
If all is well, I presume we could scrape the run history and get the properties below to look for proof of a overall job speedup

cruise_timestamp_01_scheduled | 2017-08-23T16:16:39+02:00 |  
cruise_timestamp_02_assigned | 2017-08-23T16:16:44+02:00 |  
cruise_timestamp_03_preparing | 2017-08-23T16:16:54+02:00

https://docs.gocd.org/current/advanced_usage/properties.html

**Comment by arvindsv** [comment]
@FredrikWendt @jyotisingh and @GaneshSPatil have been trying to get rid of the problems around this, over the last couple of weeks. It is still not ready. Current status: Because of the change from the pull model (old) to the push model (websockets), some state can be mismatched between the server and the agent, causing reschedules of jobs.

We want to move over to this fully, and get rid of the old way, as well. It's tough to keep both implementations working and slows all work around this area down.

**Comment by ketan** [comment]
Closing, no activity
